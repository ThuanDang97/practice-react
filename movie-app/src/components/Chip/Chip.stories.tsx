import { Story } from '@storybook/react';
import Chip from './Chip';
import { TProps } from './Chip';

export default {
  title: 'Components/Chip',
  component: Chip,
};

const Template: Story<TProps> = (args): JSX.Element => <Chip {...args} />;

export const Basic: Story<TProps> = Template.bind({});
Basic.args = {
  chip: {
    id: 1,
    name: 'Chip',
    actived: false,
  },
};

export const Actived: Story<TProps> = Template.bind({});
Actived.args = {
  chip: {
    id: 1,
    name: 'Chip',
    actived: true,
  },
};
