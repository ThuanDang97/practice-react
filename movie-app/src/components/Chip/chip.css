.list-chip {
  display: flex;
  width: 100%;
  flex-wrap: wrap;
  gap: 10px;
}

.chip-element {
  padding: 10px 18px;
  border: black 1px solid;
  border-radius: 15px;
  text-align: center;
  cursor: pointer;
  color: black;
  line-height: 15px;
}

.chip-element:hover {
  background-color: rgba(1, 180, 228, 1);
  color: #fff;
  border-color: rgba(1, 180, 228, 1);
}

.active {
  background-color: rgba(1, 180, 228, 1);
  color: #fff;
  border-color: rgba(1, 180, 228, 1);
}
