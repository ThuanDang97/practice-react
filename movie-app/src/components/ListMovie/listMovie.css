.media {
  width: 100%;
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  align-items: flex-start;
  align-content: flex-start;
}

.column_wrapper {
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: flex-start;
  align-content: flex-start;
}

.column_wrapper .content_wrapper {
  max-width: 1400px;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  align-content: flex-start;
  padding-left: 40px;
  padding-right: 40px;
  padding-top: 30px;
  padding-bottom: 30px;
}

.title {
  width: 100%;
  margin-bottom: 20px;
}

.content {
  width: 100%;
  display: flex;
  align-items: flex-start;
  align-content: flex-start;
}

.white_column {
  padding-left: 30px;
  background: transparent;
  display: flex;
  flex-wrap: wrap;
  flex: 0 1 auto;
}

.page_wrapper {
  width: 100%;
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  column-gap: 10px;
}
