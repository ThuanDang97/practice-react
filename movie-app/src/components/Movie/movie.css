.movie {
  width: 100%;
  max-height: 600px;
}

.banner {
  max-width: 1440px;
  display: flex;
  margin: 0 auto;
}

.banner .banner_image {
  position: absolute;
  right: 0;
  height: 600px;
  width: 1520px;
}

.banner .layer {
  width: 100%;
  height: 600px;
  position: absolute;
  right: 0;
  background-image: linear-gradient(
    to right,
    rgba(0, 0, 0, 1) 150px,
    rgba(220.5, 220.5, 220.5, 0.84) 100%
  );
}

.banner .content {
  z-index: 1;
  padding: 75px 40px;
}
