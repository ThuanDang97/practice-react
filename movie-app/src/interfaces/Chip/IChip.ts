export interface IChip {
  id: number;
  name: string;
  actived: boolean;
}
