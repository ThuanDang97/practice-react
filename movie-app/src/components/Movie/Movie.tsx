import { IMovieInfo } from '../../interfaces/Movie/IMovieInfo';
import ContentMovie from './ContentMovie/ContentMovie';
import './movie.css';

type TProps = {
  movie: IMovieInfo;
};

const Movie = (props: TProps): JSX.Element => {
  const { movie } = props;

  return (
    <section className="movie">
      <div className="banner">
        <img
          src={`https://www.themoviedb.org/t/p/w1920_and_h800_multi_faces${movie?.backdrop_path}`}
          alt={movie?.original_title}
          className="banner_image"
        />
        <span className="layer"></span>
        <ContentMovie
          posterPath={movie?.poster_path!}
          homepage={movie?.homepage!}
          title={movie?.title!}
          releaseDate={movie?.release_date!}
          genres={movie?.genres!}
          runtime={movie?.runtime!}
          overview={movie?.overview!}
          scoreVoteAverage={movie?.vote_average!}
        />
      </div>
    </section>
  );
};

export default Movie;
