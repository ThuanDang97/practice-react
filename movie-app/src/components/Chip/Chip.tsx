import { memo, MouseEvent } from 'react';
import { IChip } from '../../interfaces/Chip/IChip';
import './chip.css';

export type TProps = {
  chip: IChip;
  onHandleClick: (e: MouseEvent<HTMLButtonElement>) => void;
};

const Chip = (props: TProps) => {
  const { chip, onHandleClick } = props;

  return (
    <button
      value={chip?.id}
      className={chip?.actived ? 'chip-element active' : 'chip-element'}
      onClick={onHandleClick}
    >
      {chip?.name}
    </button>
  );
};

export default memo(Chip);
