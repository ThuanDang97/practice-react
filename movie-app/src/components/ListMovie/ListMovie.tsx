import { ChangeEvent, MouseEvent } from 'react';
import { IMovieCard } from '../../interfaces/Movie/IMovieCard';
import { Button } from '../Button/Button';
import Menu from '../Menu/Menu';
import CardMovie from '../CardMovie/CardMovie';
import './listMovie.css';
import Select from '../Select/Select';
import Chip from '../Chip/Chip';
import { IChip } from '../../interfaces/Chip/IChip';
import RangerSlider from '../RangerSlider/RangerSlider';
import { DefaultRangerValue } from '../../constants/Enum';

export type TProps = {
  pageMovie: IMovieCard[];
  title: string;
  chips: IChip[];
  onDisableLoadMore: boolean;
  onDisableSearch: boolean;
  onUpdateStatus: (movie: IMovieCard) => void;
  onHandleClick: (e: MouseEvent<HTMLButtonElement>) => void;
  onHandleSort?: (option: string) => void;
  onHandleFilterGenres?: (e: MouseEvent<HTMLButtonElement>) => void;
  onHandleFilterUserScore?: (e: ChangeEvent<HTMLInputElement>) => void;
  onHandleSearch: (e: MouseEvent<HTMLButtonElement>) => void;
};

const ListMovie = (props: TProps): JSX.Element => {
  const {
    pageMovie,
    title,
    chips,
    onDisableLoadMore,
    onDisableSearch,
    onUpdateStatus,
    onHandleClick,
    onHandleSort,
    onHandleFilterGenres,
    onHandleFilterUserScore,
    onHandleSearch,
  } = props;

  return (
    <>
      <div className="media" data-testid="home-page">
        <div className="column_wrapper">
          <div className="content_wrapper">
            <div className="title">
              <h2 data-testid="header-page">{title}</h2>
            </div>
            <div className="content">
              <div>
                <Menu title="Sort">
                  <Select onHandleSort={onHandleSort!} />
                </Menu>
                <Menu title="Filter">
                  <div className="list-chip" data-testid="list-chip">
                    <h3>Genres</h3>
                    {chips?.map((chip) => (
                      <Chip
                        chip={chip}
                        onHandleClick={onHandleFilterGenres!}
                        key={chip.id}
                      />
                    ))}
                  </div>
                  <RangerSlider
                    label={'User Score'}
                    min={DefaultRangerValue.Min}
                    max={DefaultRangerValue.Max}
                    step={DefaultRangerValue.Step}
                    defaultValue={DefaultRangerValue.DefaultValue}
                    onHandleFilterUserScore={onHandleFilterUserScore!}
                  />
                </Menu>
                <div className="btn-layout">
                  <Button
                    typeButton="button"
                    radiant="contained"
                    color="primary"
                    width="w-100"
                    radius="rd-2"
                    onHandleClick={onHandleSearch}
                    isDisable={onDisableSearch}
                  >
                    Search
                  </Button>
                </div>
              </div>
              <div className="white_column">
                <section className="page_wrapper">
                  {pageMovie?.length > 0 ? (
                    pageMovie?.map((movie) => (
                      <CardMovie
                        movie={movie}
                        onUpdateStatus={() => onUpdateStatus(movie)}
                        key={movie?.id}
                      />
                    ))
                  ) : (
                    <span data-testid="message-error">
                      No items were found that match your query.
                    </span>
                  )}
                </section>
                <div className="btn-layout">
                  <Button
                    typeButton="button"
                    radiant="contained"
                    color="primary"
                    width="w-100"
                    radius="rd-1"
                    onHandleClick={onHandleClick}
                    isDisable={onDisableLoadMore}
                  >
                    Load More
                  </Button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default ListMovie;
