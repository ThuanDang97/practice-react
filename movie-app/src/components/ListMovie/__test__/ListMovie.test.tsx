import { screen, render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { IChip } from '../../../interfaces/Chip/IChip';
import { IMovieCard } from '../../../interfaces/Movie/IMovieCard';
import { setup } from '../../../support/Setup';
import ListMovie from '../ListMovie';
import { TProps } from '../ListMovie';

describe('Test List Movie component', () => {
  const mockDataListMovies: IMovieCard[] = [
    {
      adult: false,
      backdrop_path: '/wcKFYIiVDvRURrzglV9kGu7fpfY.jpg',
      genre_ids: [],
      id: 453395,
      is_favorite: true,
      original_language: 'en',
      original_title: 'Doctor Strange in the Multiverse of Madness',
      overview:
        'Doctor Strange, with the help of mystical allies both old and new, traverses the mind-bending and dangerous alternate realities of the Multiverse to confront a mysterious new adversary.',
      popularity: 7647.02,
      poster_path: '/9Gtg2DzBhmYamXBS1hKAhiwbBKS.jpg',
      release_date: '2022-05-04',
      title: 'Doctor Strange in the Multiverse of Madness',
      video: false,
      vote_average: 7.6,
      vote_count: 2915,
    },
  ];

  const mockDataListChip: IChip[] = [
    {
      id: 1,
      name: 'Action',
      actived: false,
    },
    {
      id: 2,
      name: 'Drama',
      actived: false,
    },
  ];

  const defaultProps: TProps = {
    pageMovie: mockDataListMovies,
    title: 'Popular Movies',
    chips: mockDataListChip,
    onDisableLoadMore: false,
    onDisableSearch: false,
    onUpdateStatus: jest.fn(),
    onHandleClick: jest.fn(),
    onHandleSort: jest.fn(),
    onHandleFilterGenres: jest.fn(),
    onHandleFilterUserScore: jest.fn(),
    onHandleSearch: jest.fn(),
  };

  const customProps: TProps = {
    pageMovie: [],
    title: 'Popular Movies',
    chips: mockDataListChip,
    onDisableLoadMore: false,
    onDisableSearch: false,
    onUpdateStatus: jest.fn(),
    onHandleClick: jest.fn(),
    onHandleSort: jest.fn(),
    onHandleFilterGenres: jest.fn(),
    onHandleFilterUserScore: jest.fn(),
    onHandleSearch: jest.fn(),
  };

  test('Component List Movie matches DOM Snapshot', () => {
    const component = (
      <Router>
        <ListMovie {...defaultProps} />;
      </Router>
    );
    const { container } = render(component);

    expect(container).toMatchSnapshot();
  });

  test('Component List Movie should render correctly', () => {
    setup(
      <Router>
        <ListMovie {...defaultProps} />;
      </Router>,
    );

    // rendering header must be correct
    expect(screen.getByTestId('header-page')).toHaveTextContent('Popular Movies');
    expect(screen.getByText('Genres')).toHaveTextContent('Genres');
    expect(screen.getByText('User Score')).toHaveTextContent('User Score');

    // rendering content button must be correct
    expect(screen.getByText('Search')).toHaveTextContent('Search');
    expect(screen.getByText('Load More')).toHaveTextContent('Load More');

    // rendering chip must be correct
    expect(screen.getByText('Action')).toHaveTextContent('Action');
    expect(screen.getByText('Drama')).toHaveTextContent('Drama');

    // rendering default option must be correct
    expect(screen.getByText('Popularity Descending')).toHaveTextContent(
      'Popularity Descending',
    );

    // rendering content card movie must be correct
    expect(
      screen.getByText('Doctor Strange in the Multiverse of Madness'),
    ).toHaveTextContent('Doctor Strange in the Multiverse of Madness');
    expect(screen.getByText('76')).toHaveTextContent('76');
    expect(screen.getByText('2022-05-04')).toHaveTextContent('2022-05-04');
  });

  test('Component List Movie should render message when movie list is empty', () => {
    setup(
      <Router>
        <ListMovie {...customProps} />;
      </Router>,
    );

    expect(screen.getByTestId('message-error')).toHaveTextContent(
      'No items were found that match your query.',
    );
  });
});
