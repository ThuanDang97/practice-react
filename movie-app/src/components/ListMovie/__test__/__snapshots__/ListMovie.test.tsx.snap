// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test List Movie component Component List Movie matches DOM Snapshot 1`] = `
<div>
  <div
    class="media"
    data-testid="home-page"
  >
    <div
      class="column_wrapper"
    >
      <div
        class="content_wrapper"
      >
        <div
          class="title"
        >
          <h2
            data-testid="header-page"
          >
            Popular Movies
          </h2>
        </div>
        <div
          class="content"
        >
          <div>
            <div
              class="panel-card"
            >
              <div
                class="name"
              >
                <h2>
                  Sort
                </h2>
              </div>
              <div
                class="filter"
              >
                <span
                  class="widget"
                  data-testid="sub-menu"
                >
                  <span
                    class="dropdown"
                    data-testid="select"
                  >
                    <span
                      class="input"
                      data-testid="titleOption"
                    >
                      Popularity Descending
                    </span>
                    <span
                      class="select"
                    >
                      <span
                        class="icon-select"
                      />
                    </span>
                  </span>
                </span>
              </div>
            </div>
            <div
              class="panel-card"
            >
              <div
                class="name"
              >
                <h2>
                  Filter
                </h2>
              </div>
              <div
                class="filter"
              >
                <span
                  class="widget"
                  data-testid="sub-menu"
                >
                  <div
                    class="list-chip"
                    data-testid="list-chip"
                  >
                    <h3>
                      Genres
                    </h3>
                    <button
                      class="chip-element"
                      value="1"
                    >
                      Action
                    </button>
                    <button
                      class="chip-element"
                      value="2"
                    >
                      Drama
                    </button>
                  </div>
                  <div
                    class="ranger-slider"
                  >
                    <h3>
                      User Score
                    </h3>
                    <input
                      class="range"
                      max="10"
                      min="0"
                      step="1"
                      type="range"
                      value="10"
                    />
                    <div
                      class="range-value"
                    >
                      10
                    </div>
                  </div>
                </span>
              </div>
            </div>
            <div
              class="btn-layout"
            >
              <button
                class="btn primary contained rd-2 w-100 "
                type="button"
              >
                Search
              </button>
            </div>
          </div>
          <div
            class="white_column"
          >
            <section
              class="page_wrapper"
            >
              <div
                class="card"
              >
                <div
                  class="image"
                >
                  <div
                    class="wrapper"
                  >
                    <a
                      href="/movie/453395"
                    >
                      <img
                        alt="Doctor Strange in the Multiverse of Madness"
                        class="poster"
                        src="https://www.themoviedb.org/t/p/w220_and_h330_face//9Gtg2DzBhmYamXBS1hKAhiwbBKS.jpg"
                      />
                    </a>
                    <div
                      class="option"
                      data-testid="optional"
                    >
                      <span />
                    </div>
                  </div>
                </div>
                <div
                  class="content"
                >
                  <div
                    class="consensus tight"
                  >
                    <div
                      class="outer_ring"
                    >
                      <div
                        class="user_score_chart"
                        data-bar-color="#21d07a"
                        data-percent="78.0"
                        data-track-color="#204529"
                      >
                        <div
                          class="percent"
                        >
                          <span
                            class="icon"
                            data-testid="score"
                          >
                            76
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                  <h2>
                    <a
                      href="/movie/453395"
                      title="Doctor Strange in the Multiverse of Madness"
                    >
                      Doctor Strange in the Multiverse of Madness
                    </a>
                  </h2>
                  <p
                    data-testid="release-date"
                  >
                    2022-05-04
                  </p>
                </div>
              </div>
            </section>
            <div
              class="btn-layout"
            >
              <button
                class="btn primary contained rd-1 w-100 "
                type="button"
              >
                Load More
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  ;
</div>
`;
